<a href="https://www.kaggle.com/code/aayushsin7a/boston-crime?scriptVersionId=157343037" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

**Boston Crime Dataset is a collection of crime incident reports that are provided by the boston police department (BPD).**
- This dataset documents the initial details surrounding an incident to which BPD officers respond
- This is a dataset containing records from the new crime incident report system, which includes a reduced set of fields focused on capturing the type of incident as well as when and where it occured.
- Records in the new system begin from June 2015

# Data Dictionary
- INCIDENT_NUMBER -> Internal BPD report number 
- OFFENSE_CODE -> Numerical code of offense description
- OFFENSE_CODE_GROUP -> Internal categorization of offense description
- OFFENSE_DESCRIPTION -> The primary description of the incident
- DISTRICT - what district crime was reported in
- REPORTING_AREA -> RA number associated with the location where the crime was reported from
- SHOOTING ->  Indicated, a shooting took place
- OCCURRED_ON_DATE -> Earliest date and time the incident could have taken place
- UCR_Part -> Universal Crime Reporting Part Number (1,2,3)
- STREET -> Street name the incident took place
- Lat -> Latitude
- Long -> Longitude
- Location -> Location

# OSEMN Framework 
- All ML projects and Data Science projects have a basic framework - OSEMN 
- Obtaining - Data obtaining
- Scrubbing - Cleaning data for exploration
- Exploring
- Modeling
- Interpreting 

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/crimes-in-boston/crime.csv
/kaggle/input/crimes-in-boston/offense_codes.csv


In [2]:
crime_data = pd.read_csv("/kaggle/input/crimes-in-boston/crime.csv",encoding='windows-1254')
offense_code = pd.read_csv("/kaggle/input/crimes-in-boston/offense_codes.csv",encoding='windows-1254')

# Merge Dataset 


In [3]:
bostonCrime = crime_data.merge(offense_code,left_on='OFFENSE_CODE',right_on='CODE',how='left').drop('CODE',axis='columns')



In [4]:
bostonCrime.describe()

OFFENSE_CODE           YEAR          MONTH           HOUR  \
count  577880.000000  577880.000000  577880.000000  577880.000000   
mean     2231.115386    2016.562861       6.609056      13.120343   
std      1173.109538       0.996611       3.270555       6.304551   
min       111.000000    2015.000000       1.000000       0.000000   
25%       802.000000    2016.000000       4.000000       9.000000   
50%      2905.000000    2017.000000       7.000000      14.000000   
75%      3125.000000    2017.000000       9.000000      18.000000   
max      3831.000000    2018.000000      12.000000      23.000000   

                 Lat           Long  
count  545332.000000  545332.000000  
mean       42.214124     -70.907377  
std         2.164181       3.500736  
min        -1.000000     -71.178674  
25%        42.297555     -71.096744  
50%        42.325937     -71.077405  
75%        42.348624     -71.062367  
max        42.395042      -1.000000

# Finding Data Types

In [5]:
bostonCrime.dtypes

INCIDENT_NUMBER         object
OFFENSE_CODE             int64
OFFENSE_CODE_GROUP      object
OFFENSE_DESCRIPTION     object
DISTRICT                object
REPORTING_AREA          object
SHOOTING                object
OCCURRED_ON_DATE        object
YEAR                     int64
MONTH                    int64
DAY_OF_WEEK             object
HOUR                     int64
UCR_PART                object
STREET                  object
Lat                    float64
Long                   float64
Location                object
NAME                    object
dtype: object

In [6]:
print(f"Number of rows & columns: {bostonCrime.shape}")
print(f"Number of Null/Missing Values : {bostonCrime.isna().sum().values.sum()}")
print(f"Missing % in each column : { bostonCrime.isna().mean()*100}")

print("Dupication % :",
      100 * (len(bostonCrime) - 
             len(bostonCrime.drop_duplicates()))/len(bostonCrime),'%')


Number of rows & columns: (577880, 18)
Number of Null/Missing Values : 661575
Missing % in each column : INCIDENT_NUMBER         0.000000
OFFENSE_CODE            0.000000
OFFENSE_CODE_GROUP      0.000000
OFFENSE_DESCRIPTION     0.000000
DISTRICT                0.505987
REPORTING_AREA          0.000000
SHOOTING               99.657541
OCCURRED_ON_DATE        0.000000
YEAR                    0.000000
MONTH                   0.000000
DAY_OF_WEEK             0.000000
HOUR                    0.000000
UCR_PART                0.016266
STREET                  3.038693
Lat                     5.632311
Long                    5.632311
Location                0.000000
NAME                    0.000000
dtype: float64
Dupication % : 27.97380078909116 %


- Shooting has maximum % of missing values 
- Overall the dataset has 27.97 % of Missing values